### PR TITLE
Updated sample code with /scoutingObservations and /scoutringObservat…

### PR DESCRIPTION
**Pull request by rajindersaini**
…ion/{id} end points

### CODE REVIEW

File: climate.py [hunk 1]
```diff
@@ -258,3 +258,64 @@ def get_upload_status(upload_id, token, api_key):
...
+    :param occurredAfter: Optional start time by which to filter layer results.
+    :param occurredBefor: Optional end time by which to filter layer results.
+    :return: status json object containing scouting observation list and status.
+    """
+    uri = '{}/v4/layers/scoutingObservations'.format(api_uri)
+    headers = {
+        'authorization': bearer_token(token),
+        'accept': json_content_type,
+        'x-api-key': api_key,
+        'X-Limit': str(x_limit),
```
Comment (mtkp) [inline climate.py]: let's keep all headers styled the same. so far we've used lower case. so, `'x-limit'` and `'x-next-token'`

File: climate.py [hunk 2]
```diff
@@ -258,3 +258,64 @@ def get_upload_status(upload_id, token, api_key):
         return res.json()
     else:
         return None
+
+def get_scouting_observations(token, api_key, x_limit=100, x_next_token=None, occurredAfter=None, occurredBefore=None):
```
Comment (mtkp) [inline climate.py]: at the python code level, we can still use good names and consistent style. we don't need to prefix these with `x_` nor do we need to mix in camelCase. similar to what we did for `get_fields`, we can call the token key `next_token`, and we can call the limit key just `limit`. occurred before/after keys should be `occurred_after` and `occurred_before`.

File: climate.py [hunk 3]
```diff
@@ -258,3 +258,64 @@ def get_upload_status(upload_id, token, api_key):
...
+    :return: status json object containing scouting observation list and status.
+    """
+    uri = '{}/v4/layers/scoutingObservations'.format(api_uri)
+    headers = {
+        'authorization': bearer_token(token),
+        'accept': json_content_type,
+        'x-api-key': api_key,
+        'X-Limit': str(x_limit),
+        'X-Next-Token': x_next_token,
+        'occurredAfter': occurredAfter,
```
Comment (mtkp) [inline climate.py]: `'occurredAfter'` and `'occurredBefore'` are query parameters. you can pass params in a requests.get call using the `params=...` key

File: climate.py [hunk 4]
```diff
@@ -258,3 +258,64 @@ def get_upload_status(upload_id, token, api_key):
...
+        return []
+
+def get_scouting_observation(token, api_key, scouting_observation_id):
+    """
+    Retrieve an individual scouting observation by id. Ids are retrieved via the /layers/scoutingObservations route.
+    https://dev.fieldview.com/technical-documentation/ for possible status
+    values and their meaning.
+    :param token: access_token
+    :param api_key: Provided by Climate
+    :param scoutingObservationId: Unique identifier of the Scouting Observation.
```
Comment (mtkp) [inline climate.py]: nit ... param name in docstring should probably be `scouting_observation_id`

File: main.py [hunk 5]
```diff
@@ -280,14 +283,44 @@ def render_field_link(field):
...
+@app.route('/scouting-observation/<scouting_observation_id>', methods=['GET'])
+def scouting_observation(scouting_observation_id):
+    observation = climate.get_scouting_observation(state('access_token'), CLIMATE_API_KEY, scouting_observation_id)
+   
+    return  """
+            <h1>Partner API Demo Site</h1>
+            <h2>Scouting Observation ID: {scouting_observation_id}</h2>
+            <p>{json}</p>
+            <p><a href='{observations}'>Return to Observations list</a></p>
+            """.format(scouting_observation_id=scouting_observation_id,json=json.dumps(observation), observations=url_for('scouting_observations'))
```
Comment (mtkp) [inline main.py]: some newlines in between key=value pairs might be nice

File: main.py [hunk 6]
```diff
@@ -280,14 +283,44 @@ def render_field_link(field):
...
+    html = "<h1>Partner API Demo Site</h1>"
+    if observations:
+        scouting_observations = render_ul(render_scouting_observation_link(f) for f in observations)
+        html +="<p>Your Climate Scouting Observations:{scouting_observations}</p>".format(scouting_observations=scouting_observations)
+    else:
+        html+= "<p>No Scouting Observations found!"
+
+    html+= "<p><a href='{home}'>Return home</a></p>".format(home=url_for('home'))
 
+    return html
```
Comment (mtkp) [inline main.py]: how about something like

```py
if observations:
    observations_list = render_ul(render_scouting_observation_link(o) for o in observations)
    body = "Your Climate Scouting Observations: {observations_list}".format(observations_list)
else:
    body = "No Scouting Observations found!"

return """
       <h1>Partner API Demo Site</h1>
       <p>{body}</p>
       <p><a href='{home}'>Return home</a></p>
       """.format(body=body,
                  home=url_for('home'))
```

for me personally, this pattern helps communicate what the HTML will look like without needing to read the code and following the string concatenation.

File: climate.py [hunk 7]
```diff
@@ -258,3 +258,64 @@ def get_upload_status(upload_id, token, api_key):
...
+        'accept': json_content_type,
+        'x-api-key': api_key,
+        'X-Limit': str(x_limit),
+        'X-Next-Token': x_next_token,
+        'occurredAfter': occurredAfter,
+        'occurredBefore': occurredBefore 
+    }
+
+    res = requests.get(uri, headers=headers)
+    Logger().info(to_curl(res.request))
```
Comment (mtkp) [inline climate.py]: nice. i really like these curl logs

File: main.py (hunk 6 above)
Comment (tommy) [reply main.py]: +1 to this

File: main.py [hunk 8]
```diff
@@ -280,14 +283,44 @@ def render_field_link(field):
...
     return url_for('login_redirect', _external=True)
 
+@app.route('/scouting-observation/<scouting_observation_id>', methods=['GET'])
+def scouting_observation(scouting_observation_id):
+    observation = climate.get_scouting_observation(state('access_token'), CLIMATE_API_KEY, scouting_observation_id)
+   
+    return  """
+            <h1>Partner API Demo Site</h1>
+            <h2>Scouting Observation ID: {scouting_observation_id}</h2>
+            <p>{json}</p>
```
Comment (tommy) [inline main.py]: Would be good to format the json output nicely. Probably something like:
`<p><code>{json}</code></p>`

and computing json with `json.dump(observation, indent=4)`

File: climate.py (hunk 1 above)
Comment (rajindersaini) [reply climate.py]: Accepted.

File: climate.py (hunk 2 above)
Comment (rajindersaini) [reply climate.py]: Accepted.

File: climate.py (hunk 3 above)
Comment (rajindersaini) [reply climate.py]: Accepted.

File: climate.py (hunk 4 above)
Comment (rajindersaini) [reply climate.py]: Accepted.

File: main.py (hunk 5 above)
Comment (rajindersaini) [reply main.py]: Accepted

File: main.py (hunk 6 above)
Comment (rajindersaini) [reply main.py]: Accepted

File: climate.py [hunk 9]
```diff
@@ -258,3 +258,66 @@ def get_upload_status(upload_id, token, api_key):
...
+    }
+
+    res = requests.get(uri, headers=headers, params=params)
+    Logger().info(to_curl(res.request))
+
+    if res.status_code == 200:
+        return res.json()['results']
+    if res.status_code == 206:
+        next_token = res.headers['x-next-token']
+        return res.json()['results'] + get_scouting_observations(token, api_key, limit, next_token)
```
[the remark this answers is not in the document]
Comment (mtkp) [reply climate.py]: based on our offline conversation, looks like the answer is yes

File: main.py [hunk 10]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+             observations=url_for('scouting_observations'))
+
+
 
+@app.route('/scouting-observations', methods=['GET'])
+def scouting_observations():
+    observations = climate.get_scouting_observations(state('access_token'), CLIMATE_API_KEY, 100, None)
+    body = ""
+    if observations:
+        scouting_observations = render_ul(render_scouting_observation_link(f) for f in observations)
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Accepted.

File: main.py [hunk 11]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+            <p><pre>{json}</pre></p>
+            <p><a href='{observations}'>Return to Observations list</a></p>
+            """.format(scouting_observation_id=scouting_observation_id,json=json.dumps(observation, indent=4, sort_keys=True),
+             observations=url_for('scouting_observations'))
+
+
 
+@app.route('/scouting-observations', methods=['GET'])
+def scouting_observations():
+    observations = climate.get_scouting_observations(state('access_token'), CLIMATE_API_KEY, 100, None)
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Accepted.

File: main.py [hunk 12]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+@app.route('/scouting-observation/<scouting_observation_id>', methods=['GET'])
+def scouting_observation(scouting_observation_id):
+    observation = climate.get_scouting_observation(state('access_token'), CLIMATE_API_KEY, scouting_observation_id)
+   
+    return  """
+            <h1>Partner API Demo Site</h1>
+            <h2>Scouting Observation ID: {scouting_observation_id}</h2>
+            <p><pre>{json}</pre></p>
+            <p><a href='{observations}'>Return to Observations list</a></p>
+            """.format(scouting_observation_id=scouting_observation_id,json=json.dumps(observation, indent=4, sort_keys=True),
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Done

File: main.py [hunk 13]
```diff
@@ -118,7 +119,9 @@ def user_homepage():
                       fields=field_list,
                       upload=url_for('upload_form'),
                       logout=url_for('logout_redirect'),
-                      refresh=url_for('refresh_token'))
+                      refresh=url_for('refresh_token'),
+                      scouting_observations=url_for('scouting_observations'),
+                      )
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: done

File: main.py [hunk 14]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+    observations = climate.get_scouting_observations(state('access_token'), CLIMATE_API_KEY, 100, None)
+    body = ""
+    if observations:
+        scouting_observations = render_ul(render_scouting_observation_link(f) for f in observations)
+        body = "<p>Your Climate Scouting Observations:{scouting_observations}</p>".format(scouting_observations=scouting_observations)
+    else:
+        body = "<p>No Scouting Observations found!"
+
+    return """
+        <h1>Partner API Demo Site</h1>
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Done

File: main.py [hunk 15]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+            <p><a href='{observations}'>Return to Observations list</a></p>
+            """.format(scouting_observation_id=scouting_observation_id,json=json.dumps(observation, indent=4, sort_keys=True),
+             observations=url_for('scouting_observations'))
+
+
 
+@app.route('/scouting-observations', methods=['GET'])
+def scouting_observations():
+    observations = climate.get_scouting_observations(state('access_token'), CLIMATE_API_KEY, 100, None)
+    body = ""
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Done

File: main.py [hunk 16]
```diff
@@ -280,14 +283,47 @@ def render_field_link(field):
...
+    if observations:
+        scouting_observations = render_ul(render_scouting_observation_link(f) for f in observations)
+        body = "<p>Your Climate Scouting Observations:{scouting_observations}</p>".format(scouting_observations=scouting_observations)
+    else:
+        body = "<p>No Scouting Observations found!"
+
+    return """
+        <h1>Partner API Demo Site</h1>
+        <p>{body}</p>
```
[the remark this answers is not in the document]
Comment (rajindersaini) [reply main.py]: Done

File: climate.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (rajindersaini) [reply climate.py]: Yes

**Pull request merged**